File: 15-Dynamic-Programming/11-LIS-Variants/cpp/lis.cpp
/* Longest Increasing Subsequence | LeetCode 300 | O(n log n) | Tags: lis, dp */
#include <bits/stdc++.h>
using namespace std;
int lengthOfLIS(vector<int>& nums){vector<int> tails;for(int x:nums){auto it=lower_bound(tails.begin(),tails.end(),x);if(it==tails.end())tails.push_back(x);else *it=x;}return tails.size();}
int main(){vector<int>nums={10,9,2,5,3,7,101,18};cout<<lengthOfLIS(nums)<<endl;} // 4

File: 15-Dynamic-Programming/11-LIS-Variants/java/LIS.java
import java.util.*;
public class LIS { public static int lengthOfLIS(int[] nums){List<Integer> tails=new ArrayList<>();for(int x:nums){int pos=Collections.binarySearch(tails,x);if(pos<0)pos=-(pos+1);if(pos==tails.size())tails.add(x);else tails.set(pos,x);}return tails.size();}
    public static void main(String[] args){System.out.println(lengthOfLIS(new int[]{10,9,2,5,3,7,101,18}));}}
